GPflowOpt/bo.py

```diff
@@ -190,6 +222,10 @@ def inverse_acquisition(x):
 
         # Optimization loop
         for i in range(n_iter):
+            # If callback specified, and acquisition has the setup flag enabled (indicating an upcoming compilation,
+            # run the callback.
+            if self._iter_callback and self.acquisition._needs_setup:
+                self._iter_callback([m.wrapped for m in self.acquisition.models])
```

Comment:
if there is no callback:
- setup is run and models are optimized on the first evaluate
with a callback:
- models are optimized here but setup probably has not been run yet and needs_setup is still True -> models are optimized again on first evaluate? right?

Reply:
You confuse something here: you _can_ optimize your model in the callback but this is one of the scenarios (which would require optimize_restarts to be 0 in order to avoid two optimizes). The primary use case is to only set the _initial_ starting point. 

(The reason the jitchol callback runs the optimization for a small number of steps is to check if no cholesky error occurs, not to optimize the model. )

Reply:
Ok, there was indeed some confusion here. I thought the callback would implement the complete model building strategy: setting hyps, running one or more optimizations, etc. This is still possible but you have to set optimize_restarts = 0